core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java

```diff
@@ -911,7 +912,9 @@ protected ReplicaResult shardOperationOnReplica(Request request, IndexShard repl
         assertThat(capturedRequests.size(), equalTo(1));
         final CapturingTransport.CapturedRequest capturedRequest = capturedRequests.get(0);
         assertThat(capturedRequest.action, equalTo("testActionWithExceptions[r]"));
-        assertThat(capturedRequest.request, instanceOf(TransportReplicationAction.ConcreteShardRequest.class));
+        assertThat(capturedRequest.request, instanceOf(TransportReplicationAction.ConcreteReplicaRequest.class));
```

Comment:
I'm not sure if this is enough, this wouldn't reproduce the class cast exception that revealed the issue. Ideally, we would want a test that did *not* have to be changed (other than making it compile) as part of inlining global checkpoints to catch the issue.